File: 1153.string-transforms-into-another-string.java
/*
 * @lc app=leetcode id=1153 lang=java
 *
 * [1153] String Transforms Into Another String
 */

// @lc code=start

class Solution {
    public boolean canConvert(String str1, String str2) {
        if (str1.equals(str2))
            return true;
        Map<Character, Character> map = new HashMap<>();

        int len = str1.length();
        for (int i = 0; i < len; i++) {
            char ch1 = str1.charAt(i), ch2 = str2.charAt(i);
            if (map.putIfAbsent(ch1, ch2) != null && ch2 != map.putIfAbsent(ch1, ch2)) {
                return false;
            }
        }

        return new HashSet<Character>(map.values()).size() < 26;
    }
}
// @lc code=end

File: 1153.string-transforms-into-another-string.py
#
# @lc app=leetcode id=1153 lang=python3
#
# [1153] String Transforms Into Another String
#

# @lc code=start

class Solution:
    def canConvert(self, str1: str, str2: str) -> bool:
        if str1 == str2:
            return True

        convertMap = dict()
        letterSet = set()

        for char1, char2 in zip(str1, str2):
            if char1 not in convertMap:
                convertMap[char1] = char2
                letterSet.add(char2)
            elif convertMap[char1] != char2:
                return False

        # if len(letterSet) == 26,
        # "abcdefghijklmnopqrstuvwxyz" -> "bcdefghijklmnopqrstuvwxyza"
        # 1. "abcdefghijklmnopqrstuvwxyz" -> "abcdefghijklmnopqrstuvwxya"
        # 2. "abcdefghijklmnopqrstuvwxya" -> "bbcdefghijklmnopqrstuvwxyb"
        # ... impossible
        return len(letterSet) < 26


# @lc code=end
